os/arch/arm/src/armv7-a/arm_cpugating.c

```diff
@@ -40,7 +40,22 @@
 #include "barriers.h"
 
 #ifdef CONFIG_CPU_GATING
-volatile uint32_t ulFlashPG_Flag = 0;
+static volatile uint32_t g_cpugating_flag[CONFIG_SMP_NCPUS];
+
+void up_set_gating_flag_status(uint32_t CoreID, uint32_t val)
+{
+	g_cpugating_flag[CoreID] = val;
```

Comment:
This doesn't seem to fit the pair.

when flag setting is plus 1 `g_cpugating_flag[cpu]++;`,
but  flag release is set 0 foced.

Is this okay when the request comes at the same time?

Reply:
Hi eunwoo, I think there is some restriction on the usage here, and it should only be applied on some special cases (ie. we need to freeze the state of some specific cores).
However, when a gating request is sent, the requestor normally should do a polling to check the status of the gated cpu. And the freeing of flag should be called after some checking/operation is completed. In this case, it should work as a pair.
So, I think some assumption or limitation should be explained more clearly if someone is applying this.

Reply:
And does gating mean forced getting the WFE state?

I couldn't find cpu gating, I just found power gating of core.

Reply:
> It doesn't seem necessary for the CPU gating code to know about flags. Wouldn't a function like up_cpu_gating_release() be needed?

What is the benefit of doing so? In SGI2, there is also a up_cpu_pausereq to get the flag of cpu. I don't think we need another API to do that.

> And if there are more secondary CPUs, it seems that the release of this state should only occur when all cores have been released. Is there any reason why the flag needs to be initialized to zero by force?

 I don't understand what you are referring to.

Reply:
> And does gating mean forced getting the WFE state?
> 
> I couldn't find cpu gating, I just found power gating of core.

CPU gating means, the target CPU is blocked to execute any operation, thus we use an infinite loop to force blocking the CPU. WFE is applied to inhibit additional power consumption.

Reply:
> > It doesn't seem necessary for the CPU gating code to know about flags. Wouldn't a function like up_cpu_gating_release() be needed?
> 
> What is the benefit of doing so? In SGI2, there is also a up_cpu_pausereq to get the flag of cpu. I don't think we need another API to do that.

up_cpu_pausereq is using the SP_LCOK and limit this flag in arm_cpupause. It dosen't allow setting by other code.
If you should use flag, you need to limit in this file.
If you allow set flag using up_set_gating_flag_status, we can't avoid wrong used it form other module. 

> > And if there are more secondary CPUs, it seems that the release of this state should only occur when all cores have been released. Is there any reason why the flag needs to be initialized to zero by force?
> 
> I don't understand what you are referring to.

`g_cpugating_flag[cpu]++;` is counting structure. 
but up_set_gating_flag_status(cpu, 0) is calling with 0 or 1 argument.

I thought this was for a 4-core case.
It's because all CPUs have to maintain gating until it is released for gating. However, this seems to be handled by using duplicate delay interrupts.
Do we need to have a counting structure right?

Reply:
> up_cpu_pausereq is using the SP_LCOK and limit this flag in arm_cpupause. It dosen't allow setting by other code.
> If you should use flag, you need to limit in this file.
> If you allow set flag using up_set_gating_flag_status, we can't avoid wrong used it form other module.

Understood. But, why would we use it wrongly? This should only be used under specific circumstances.

> g_cpugating_flag[cpu]++; is counting structure.
> but up_set_gating_flag_status(cpu, 0) is calling with 0 or 1 argument.
> 
> I thought this was for a 4-core case.
> It's because all CPUs have to maintain gating until it is released for gating. However, this seems to be handled by using duplicate delay interrupts.
> Do we need to have a counting structure right?

Yes, this supports up to 4 cores case. The structure follows CONFIG_SMP_NCPUS. 0 or 1 argument is set flag to control CPU state. What do you mean by duplicate delay interrupts? Currently, we gate each core and check its condition one by one, and if one of the secondary CPU violate the condition, we release gated core in descending order. Is there any misunderstanding...?

Reply:
> > up_cpu_pausereq is using the SP_LCOK and limit this flag in arm_cpupause. It dosen't allow setting by other code.
> > If you should use flag, you need to limit in this file.
> > If you allow set flag using up_set_gating_flag_status, we can't avoid wrong used it form other module.
> 
> Understood. But, why would we use it wrongly? This should only be used under specific circumstances.
> 
> > g_cpugating_flag[cpu]++; is counting structure.
> > but up_set_gating_flag_status(cpu, 0) is calling with 0 or 1 argument.
> > I thought this was for a 4-core case.
> > It's because all CPUs have to maintain gating until it is released for gating. However, this seems to be handled by using duplicate delay interrupts.
> > Do we need to have a counting structure right?
> 
> Yes, this supports up to 4 cores case. The structure follows CONFIG_SMP_NCPUS. 0 or 1 argument is set flag to control CPU state. What do you mean by duplicate delay interrupts? Currently, we gate each core and check its condition one by one, and if one of the secondary CPU violate the condition, we release gated core in descending order. Is there any misunderstanding...?

`we release gated core in descending order.`
Could you let me know where code release gating count?
current code, only set 1 and 0....

Reply:
@ewoodev , Please check the code here https://github.com/Samsung/TizenRT/pull/6256/files#diff-34272b827c1d9af918fa7e69ffe694e7e31445afb6d3e04081179ea1a2bb5225R168-R174.
If one of the secondary cores cannot go hot plug, we set the flag to 0 and do SEV.
And then, https://github.com/Samsung/TizenRT/pull/6256/files#diff-af4e58c133b5aca9d3f92e02b429cae7c5ad60207bfee9e537c5cb97ebb9ef2cR80-R81. Once the condition does not meet, it will be released from gated state and resume its operation.

Reply:
@edwakuwaku Thank you for explain detail.
I still couldn't understand this.

You call up_cpu_hotplug() for CPU gating. However, to release the gating, you directly set the flag using up_set_gating_flag_status().

There is no setting of a flag for CPU gating, but for release gating we need to set flag manually. I think this doesn't match general idea of a lock/unlock pair.

Secondly, for CPU gating, you increment g_cpugating_flag with ++. But when releasing the gating, you set it to 0. I can't understand why there's a counting structure instead of a boolean(set 0, 1) If we did ++, then we should do --.

Could you let me know your opinion?

Reply:
Hello @ewoodev , I understand your concern now.
The reason I added the g_cpugating_flag increment, is to ensure that the gating will be success in any circumstances (ie. if any user invoke gating SGI, but flag is not set).
Anyway, your point make sense to me. We could add setting flag to 1 when invoke up_cpu_gating. In that case, it would be a pair.
Also as you mentioned, if there is other usage applying this, then a counting structure is necessary to differentiate that as well. But, if any user need to apply this, it has to be used with care, if synchronization is not done carefully, it would cause hang issue easily.